Adding info handler, that contains two fields, num objects and memory…

zackcam opened the pull request:
Added an info handler which required updating Cargo.toml to include valkey-module-macros and linkme. The info handler contains two fields bloom_total_memory_bytes and bloom_num_objects. I then updated the creation and copy methods of bloom objects to increment these values where necessary. In order to deincrement added impl Drop for bloom in order to decrease the info fields as necessary.

Testing:
Created a new test file specifically for info. This test file contains for different tests:
test_basic_command_metrics: This tests all create commands and non create commands to check that create commands increase the info metrics while non create dont have an influence.
test_scaled_bloomfilter_metrics: Tests scaling a bloom filter increases the metrics as desired
test_copy_metrics: Tests copying a bloom filter will double the metrics as desired
test_save_and_restore_metrics: Tests saving and restoring a scaled and non scaled bloom filter will increase metrics as desired

KarthikSubbarao commented on src/configs.rs:
```diff
@@ -26,6 +26,8 @@ lazy_static! {
     pub static ref BLOOM_EXPANSION: AtomicI64 = AtomicI64::new(BLOOM_EXPANSION_DEFAULT);
     pub static ref BLOOM_MEMORY_LIMIT_PER_FILTER: AtomicI64 =
         AtomicI64::new(BLOOM_MEMORY_LIMIT_PER_FILTER_DEFAULT);
+    pub static ref BLOOM_NUM_OBJECTS: AtomicUsize = AtomicUsize::new(0);
+    pub static ref BLOOM_OBJECT_TOTAL_MEMORY_BYTES: AtomicUsize = AtomicUsize::new(0);
```
This works. But for clarity, can we create a `src/metrics.rs` file? We can add both the lazy static and the info handler macro there

KarthikSubbarao commented on src/bloom/utils.rs:
```diff
@@ -199,11 +207,15 @@ impl BloomFilter {
             capacity,
-        }
+        };
+        configs::BLOOM_NUM_OBJECTS.fetch_add(1, Ordering::Relaxed);
```
BLOOM_NUM_OBJECTS is the number of bloom objects. 

We can add a different metric, `BLOOM_NUM_FILTERS`, to track the number of filters created as a result of existing bloom objects

KarthikSubbarao commented on src/bloom/utils.rs:
```diff
@@ -56,6 +56,10 @@ impl BloomFilterType {
         expansion: u32,
         validate_size_limit: bool,
     ) -> Result<BloomFilterType, BloomError> {
+        configs::BLOOM_OBJECT_TOTAL_MEMORY_BYTES.fetch_add(
```
We should increment `BLOOM_NUM_OBJECTS ` here, right?

YueTang-Vanessa replied:
Why increase this metric here? What if there has error? Maybe consider to move this few lines below?

KarthikSubbarao commented on src/bloom/utils.rs:
```diff
@@ -221,11 +233,15 @@ impl BloomFilter {
             capacity,
-        }
+        };
+        configs::BLOOM_NUM_OBJECTS.fetch_add(1, Ordering::Relaxed);
```
We should not increment `BLOOM_NUM_OBJECTS` here. But, instead increment `BLOOM_NUM_FILTERS`.

You can first update all the right places for these two metrics. I will review the next commit where these are fixed

YueTang-Vanessa commented on src/bloom/data_type.rs:
```diff
@@ -108,6 +110,10 @@ impl ValkeyDataType for BloomFilterType {
             filters.push(filter);
         }
+        BLOOM_OBJECT_TOTAL_MEMORY_BYTES.fetch_add(
+            mem::size_of::<BloomFilterType>(),
```
Will there has duplicate memory add here as few lines above BloomFilter::from_existing will also increase same metric `BLOOM_OBJECT_TOTAL_MEMORY_BYTES`?

zackcam replied:
I don't think so this never has from_exisitng or new called for BloomFilterType as far as I could see. This should also be caught in my rdb test if it would double up. I think its confusing because there is a from_existing in both BloomFilter and BloomFilterType. Does this make more sense now?

zackcam commented:
Accidently closed the pr, reopening with the following changes:

1. Moved the info handler to a separate file as well as the info metrics.
2. Created a new metric for number of filters (Changed the other metric to actually be number of bloom objects)
3. Updated tests to include this new metric
4. Moved an increment of metrics to be after a possible thrown error

YueTang-Vanessa commented on tests/test_bloom_metrics.py:
```diff
@@ -0,0 +1,154 @@
+
+
+        assert total_memory_bites == expected_memory 
+        assert num_objects == expected_num_objects
```
Missing `assert num_filters == expected_num_filters`

YueTang-Vanessa commented on tests/test_bloom_metrics.py:
```diff
@@ -0,0 +1,154 @@
+                num_objects = int(line.split(':')[1])
+            elif line.startswith('bf_bloom_num_filters'):
+                num_filters = int(line.split(':')[1])
+
+
```
Nit pick: may just keep one newline

YueTang-Vanessa commented on tests/test_bloom_metrics.py:
```diff
@@ -0,0 +1,154 @@
+    '''
+    Test data set unit basic metrics
+    '''
+    def test_basic_command_metrics(self):
```
Can you also add one test case of key expiration? Use `EXPIRE` for a bloom filter key, and see if all metrics works as expected.

zackcam replied:
Added this case in test_basic_command_metrics

YueTang-Vanessa reviewed (approved):
LGTM

KarthikSubbarao commented on src/metrics.rs:
```diff
@@ -0,0 +1,32 @@
+            BLOOM_NUM_OBJECTS.load(Ordering::Relaxed).to_string(),
+        )?
+        .field(
+            "bloom_num_filters",
```
`bloom_num_filters_across_objects`

KarthikSubbarao commented on src/metrics.rs:
```diff
@@ -0,0 +1,32 @@
+use valkey_module::{InfoContext, ValkeyResult};
+
+lazy_static! {
+    pub static ref BLOOM_NUM_OBJECTS: AtomicUsize = AtomicUsize::new(0);
```
Can we have this as a u64?

KarthikSubbarao commented on src/metrics.rs:
```diff
@@ -0,0 +1,32 @@
+lazy_static! {
+    pub static ref BLOOM_NUM_OBJECTS: AtomicUsize = AtomicUsize::new(0);
+    pub static ref BLOOM_OBJECT_TOTAL_MEMORY_BYTES: AtomicUsize = AtomicUsize::new(0);
+    pub static ref BLOOM_NUM_FILTERS: AtomicUsize = AtomicUsize::new(0);
```
Can we have this as a u64?

KarthikSubbarao commented on tests/test_bloom_metrics.py:
```diff
@@ -0,0 +1,166 @@
+    # expected_memory: the size of the memory used by the objects
+	# expected_num_objects: the number of module objects stored
+    # expected_num_filters: the number of filters currently created
+    def verify_bloom_metrics(self, info_response, expected_memory, expected_num_objects, expected_num_filters):
```
Can we move this into the main valkey bloom test class?

zackcam replied:
Moved and comment is updated

KarthikSubbarao commented on src/lib.rs:
```diff
@@ -57,6 +60,14 @@ fn bloom_insert_command(ctx: &Context, args: Vec<ValkeyString>) -> ValkeyResult
+///
+#[info_command_handler]
+fn add_info(ctx: &InfoContext, _for_crash_report: bool) -> ValkeyResult<()> {
+    create_info_handler(ctx)
```
Nit: A better name could be: `bloom_info_handler`

This fits well as if we add other data types, we can add more x_info_handlers to this function

KarthikSubbarao commented on src/lib.rs:
```diff
@@ -57,6 +60,14 @@ fn bloom_insert_command(ctx: &Context, args: Vec<ValkeyString>) -> ValkeyResult
+/// Module Info
+///
+#[info_command_handler]
+fn add_info(ctx: &InfoContext, _for_crash_report: bool) -> ValkeyResult<()> {
```
Nit: A better name could be: `info_handler`
